Save empty style sheet when saving web page resources

szewai opened the pull request:
#### 99a6e245f9ebe513bc7edc7055594b9436bc4472
<pre>
Save empty style sheet when saving web page resources
<a href="https://bugs.webkit.org/show_bug.cgi?id=270711">https://bugs.webkit.org/show_bug.cgi?id=270711</a>
<a href="https://rdar.apple.com/121205393">rdar://121205393</a>

Reviewed by Chris Dumez.

We currently replace URLs of empty style sheets with relative file paths when saving page resources, but we don&apos;t
actually create files for them (as the content is empty), so the saved web page will point to a file that does not
exist. To fix this, ensure we create an file on disk even the content is empty.

* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::addSubresourcesForCSSStyleSheetsIfNecessary):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: <a href="https://commits.webkit.org/275935@main">https://commits.webkit.org/275935@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/89c939e1fbede222b31b6cdf943293b2e3909c5a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43332 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22362 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45742 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45967 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39458 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45636 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26112 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19780 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35824 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43906 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19382 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/37324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/16804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/16964 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/38382 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1388 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39513 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/38701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47511 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/18246 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/15015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/42607 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19783 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41264 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9636 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/19962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19414 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:
```diff
@@ -1385,6 +1385,73 @@ function onImageLoad() {
+
+TEST(WebArchive, SaveResourcesEmptyStyleSheet)
+{
+    RetainPtr<NSURL> directoryURL = [NSURL fileURLWithPath:[NSTemporaryDirectory() stringByAppendingPathComponent:@"SaveResourcesTest"] isDirectory:YES];
```
Can probably use `RetainPtr`

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:
```diff
@@ -1385,6 +1385,73 @@ function onImageLoad() {
+    NSFileManager *fileManager = [NSFileManager defaultManager];
+    [fileManager removeItemAtURL:directoryURL.get() error:nil];
+
+    auto configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
```
`RetainPtr`

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:
```diff
@@ -1385,6 +1385,73 @@ function onImageLoad() {
+    [fileManager removeItemAtURL:directoryURL.get() error:nil];
+
+    auto configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
+    auto schemeHandler = adoptNS([[TestURLSchemeHandler alloc] init]);
```
`RetainPtr`

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:
```diff
@@ -1385,6 +1385,73 @@ function onImageLoad() {
+    auto schemeHandler = adoptNS([[TestURLSchemeHandler alloc] init]);
+    [configuration setURLSchemeHandler:schemeHandler.get() forURLScheme:@"webarchivetest"];
+    NSData *htmlData = [NSData dataWithBytes:htmlDataBytesForEmptyStyleSheet length:strlen(htmlDataBytesForEmptyStyleSheet)];
+    NSData *cssData = [[NSData alloc] init];
```
Seems like we should adopt `[[NSData alloc] init]`, no? Seems like this may be leaking.

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:
```diff
@@ -1385,6 +1385,73 @@ function onImageLoad() {
+            data = cssData;
+        }
+
+        auto response = adoptNS([[NSURLResponse alloc] initWithURL:task.request.URL MIMEType:mimeType expectedContentLength:data.length textEncodingName:nil]);
```
`RetainPtr`

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:
```diff
@@ -1385,6 +1385,73 @@ function onImageLoad() {
+        [task didFinish];
+    }];
+
+    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:NSMakeRect(0, 0, 800, 600) configuration:configuration.get()]);
```
`RetainPtr`

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:
```diff
@@ -1385,6 +1385,73 @@ function onImageLoad() {
+        EXPECT_NULL(error);
+        NSString *mainResourcePath = [directoryURL URLByAppendingPathComponent:@"host.html"].path;
+        EXPECT_TRUE([fileManager fileExistsAtPath:mainResourcePath]);
+        NSString *savedMainResource = [[NSString alloc] initWithData:[NSData dataWithContentsOfFile:mainResourcePath] encoding:NSUTF8StringEncoding];
```
Missing `adoptNS()`?

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:
```diff
@@ -1385,6 +1385,73 @@ function onImageLoad() {
+
+        NSString *resourceDirectoryName = @"host_files";
+        NSString *styleReourceName = @"empty.css";
+        NSString *resourceDirectoryPath =[directoryURL URLByAppendingPathComponent:resourceDirectoryName].path;
```
missing space after `=`

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:
```diff
@@ -1385,6 +1385,73 @@ function onImageLoad() {
+        NSString *resourceDirectoryPath =[directoryURL URLByAppendingPathComponent:resourceDirectoryName].path;
+        NSString *styleResourceFilePath = [resourceDirectoryPath stringByAppendingPathComponent:styleReourceName];
+        EXPECT_TRUE([fileManager fileExistsAtPath:styleResourceFilePath]);
+        NSString *savedStyleResource = [[NSString alloc] initWithData:[NSData dataWithContentsOfFile:styleResourceFilePath] encoding:NSUTF8StringEncoding];
```
Missing `adoptNS()`?

cdumez reviewed (approved):
LGTM but I think there are a few issues in the test.

szewai committed:
Save empty style sheet when saving web page resources
https://bugs.webkit.org/show_bug.cgi?id=270711
rdar://121205393

Reviewed by Chris Dumez.

We currently replace URLs of empty style sheets with relative file paths when saving page resources, but we don't
actually create files for them (as the content is empty), so the saved web page will point to a file that does not
exist. To fix this, ensure we create an file on disk even the content is empty.

* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::addSubresourcesForCSSStyleSheetsIfNecessary):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: https://commits.webkit.org/275935@main
